Generate publication ready oncoprint landscape figure

cbethell opened the pull request:
### Purpose/implementation Section

The purpose of this PR is to add a script to make the oncoprint landscape figure publication ready as depicted in the sketch [here](https://github.com/AlexsLemonade/OpenPBTA-manuscript/blob/master/content/images/figure_sketches/oncoprint-proposed-sketch.png).

#### What scientific question is your analysis addressing?

This PR pulls out the necessary results and steps required to rerun the oncoprint figure panel plots when necessary.

#### What was your approach?

- I took the bulk of the main functionality steps in `01-plot-oncoprint.R` and compiled them into a function script found at `util/oncoplot-functions.R`
- I then sourced this script in `01-plot-oncoprint.R` and in `figures/fig3-oncoprint-landscape.R`
- The `figures/fig3-oncoprint-landscape.R` script reruns the primary-plus (primary and secondary specimens) and primary-only versions of the oncoprint that take in the genes of interest list.

- I also adjusted the `run-oncoprint.sh` script from the `oncoprint-landscape` directory to take in the consensus seg autosomes file instead of the controlfreec autosomes file and removed the steps relevant to the gene lists that were previously being implemented.

#### What GitHub issue does your pull request address?

This PR addresses issue #641.

### Directions for reviewers. Tell potential reviewers what kind of feedback you are soliciting.

#### Which areas should receive a particularly close look?

- Does this appear to be the most efficient manner for re-running the oncoprint landscape plots?
- I was unsuccessful in my attempt to use `patchwork` to assemble the final figure due to what I believe is just the functionality of the `oncoplot` function.
It does not allow me to save the individual plots as an object in the R environment, therefore preventing me from using the `patchwork::plot_layout()`function.
**Are there any suggestions for an alternative here?**
- I also added the short histology names and hex codes to the `oncoplot_palette.R` script so that they can be sourced along with the remaining hex codes that are relevant to the oncoprint.
However, I am not sure that this was the best decision so this may need a close look as well.

#### Is there anything that you want to discuss further?

**Note: The generation of the oncoplot step causes the R session to crash on my local machine.
Therefore, the current results in this PR were produced using the testing subset data.**

Note also that:
- the color palettes found in `figures/palettes` still need to be incorporated
- the copy number calls will be adjusted upon completion of PR #644 and this may affect the figure generation script
- no plots are included in this PR yet (due to the session crashing locally)

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?

Yes, although a method to assembly the individual plots into a final figure is still left to be determined.

### Results

#### What types of results are included (e.g., table, figure)?

- The four oncoprint plots in `analyses/oncoprint-landscape/plots` have been updated using the consensus seg autosomes file, the gene list from the `interaction-plots/results` directory, and the additional hex codes in the updated `oncoplot-palette.R`

- There are also two plots saved in `figures/png` named `fig3a-oncoprint-landscape.png` and `fig3b-oncoprint-landscape.png`.
As mentioned above, these represent the primary-plus and primary-only oncoprints respectively, both filtered to represent only the genes of interest.
**These plots are here temporarily until a solution for assembling them into one final figure is determined.**

**Note also: In an attempt to find a solution for assembling these plots into one final figure, I implemented the `coOncoplot` function in the fig3 script and saved the output at `figures/pngs/fig3-oncoprint-landscape.png`.
This format is not ideal as it places the oncoprints side by side but it is a possible solution.**


#### What is your summary of the results?


#### Reproducibility Checklist

<!-- Check all those that apply or remove this section if it is not applicable.-->

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [ ] This analysis has been added to continuous integration.
I have not yet added the figure script to CI as I have not successfully ran it locally.

#### Documentation Checklist

<!-- Please review https://github.com/AlexsLemonade/OpenPBTA-analysis#documenting-your-analysis -->

- [ ] This analysis module has a `README` and it is up to date.
- [ ] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

<!-- IF YOUR PULL REQUEST IS A DATA RELEASE, PLEASE REMOVE THE [HTML COMMENT TAG](https://html.com/tags/comment-tag/) FROM THE SECTION BELOW AND COMPLETE THE CHECKLIST-->

<!-- 
### Data Release Checklist
- [ ] Is the table in doc/data-file-descriptions.md up to date?
- [ ] Is doc/data-format.md up to date?
- [ ] Is doc/release-notes.md up to date?
- [ ] Is download-data.sh up to date?
-->

cansavvy commented on figures/scripts/fig3-oncoprint-landscape.R:
```diff
@@ -0,0 +1,240 @@
+
+#### Custom function ----------------------------------------------------------
+
+prepare_and_plot_oncoprint <- function(maf_df,
```
How is this section different or the same from the `01-plot-oncoprint.R`?

cansavvy replied:
Can you put a link to  where this code originally comes from? Also is there a way we can minimize the duplication of the same code? (I get that we want to render the final png in one place so the answer to this may be no. )

cansavvy replied:
Another idea to be thinking about is if you could make this function into its own `util` file in oncoprint-landscape folder, and then source it. Use it in `oncoprint-landscape` and here. (I did a similar thing for the CDF plotting function if you want to look for an example of that. This makes it so you only have to change this function once in the future and it will be changed in both the final figure and the figures in the `oncoprint-landscape`.

cbethell replied:
> How is this section different or the same from the 01-plot-oncoprint.R?

This section in particular is the same. After making this script I did wonder if there was a way for me to take the two output plots needed from `01-plot-oncoprint.R` and read them into a script to be assembled into a figure panel but have not determined a way to do this as yet.

That being said, your suggestion:
> Another idea to be thinking about is if you could make this function into its own util file in oncoprint-landscape folder, and then source it. Use it in oncoprint-landscape and here. (I did a similar thing for the CDF plotting function if you want to look for an example of that. This makes it so you only have to change this function once in the future and it will be changed in both the final figure and the figures in the oncoprint-landscape.

seems that it will likely be the best solution in this situation.

cbethell commented on analyses/oncoprint-landscape/run-oncoprint.sh:
```diff
@@ -15,26 +15,28 @@ cd "$script_directory" || exit
-mkdir -p driver-lists
-wget -N --directory-prefix=driver-lists -O driver-lists/brain-goi-list-long.txt https://github.com/marislab/create-pptc-pdx-oncoprints/raw/2c9ed2a2331bcef3003d6aa25a130485d76a3535/data/brain-goi-list-old.txt
-wget -N --directory-prefix=driver-lists -O driver-lists/brain-goi-list-short.txt https://github.com/marislab/create-pptc-pdx-oncoprints/raw/2c9ed2a2331bcef3003d6aa25a130485d76a3535/data/brain-goi-list.txt
+# mkdir -p driver-lists
```
I commented it out for development reasons mainly. We are using the gene list from the `interaction-plots` directory in this PR instead of these driver gene lists so this is technically not needed here unless we want to leave the option in to implement them instead.

cbethell commented on figures/generate-figures.sh:
```diff
@@ -48,11 +50,33 @@ bash ${analyses_dir}/interaction-plots/01-create-interaction-plots.sh
+######################
+## Oncoprint plot(s)
 
+# Primary only oncoprint
```
This distinguished in the `independent-samples` module [here](https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/79a698904bb5f2e73c9eeef2918f48c80d6323e7/analyses/independent-samples/README.md#summary).

cbethell commented on figures/scripts/fig3-oncoprint-landscape.R:
```diff
@@ -0,0 +1,240 @@
+output_dir <- file.path(root_dir, "figures", "pngs")
+
+# Source the color palette for plots
+source(
```
That's a good idea. However, these colors would also need to be further tested and approved as I do not believe they were before.

cansavvy replied:
Can you put a short description of this color palette file in https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/510?

cansavvy replied:
I am not 100% sure what we will want to do with it, but I think it needs to be in the README at least, or needs to be incorporated into another color palette strategy at most. But if we have it recorded there we can see what else pops up as we are working on these figures.

cansavvy commented on analyses/oncoprint-landscape/util/oncoplot-functions.R:
```diff
@@ -0,0 +1,103 @@
+  
+  # If this function is being run in the figures generation script, return the
+  # maf object (instead of the final plot)
+  if (running_in_figures_script) {
```
Can you explain this distinction? Why are you plotting it in one situation and not in the other? Is there a way we can revise `analyses/oncoprint-landscape/01-plot-oncoprint.R` so we have less of a branch off? Secondly, I would have expected this function to always plot based on the name `prepare_and_plot_oncoprint`.

cbethell replied:
The issue here is that the `coOncoplot` function (that I used to lay out the two specific oncoprints side by side in the figure generation script expects two maf objects, while the `oncoplot` function used to produce each individual oncoprint in the analyses/oncoprint-landscape/plots module only expects one maf object.

Should I maybe add the coOncoplot step to this function for `running_in_figures_script` = `TRUE` and add the function argument `maf_df_2` = `NULL` for the purpose of consistency?

cansavvy replied:
Do you end up using `coOncoplot` in both scenarios? If yes, then, let's find a way to incorporate it into the `prepare_and_plot_oncoprint ` function, if no, then we might want to remove it from this particular function.

cbethell replied:
No, should I should remove the `oncoplot` step from this function and put it in the `01-plot-oncoprint.R` script in that case?

cansavvy replied:
Upon digging into the code a bit I would say, yes. Put the `oncoprint` step into `01-plot-oncoprint.R` and remove it from `prepare_and_plot_oncoprint`. Then you will should probably: 

1) Get rid of the `running_in_figures_script` argument. 
2) Rename `prepare_and_plot_oncoprint` it something that indicates it won't do the plotting, but is preparing it for the plotting. `prepare_for_oncoprint` or something like that.

cbethell commented on analyses/oncoprint-landscape/util/oncoplot-functions.R:
```diff
@@ -0,0 +1,103 @@
+  
+  if (!is.null(fusion_df)) {
+    # Bind rows of maf and fusion data frames
+    maf_df <- dplyr::bind_rows(maf_df, fusion_df)
```
Yes it does create NAs in the data frame. `fusion_df` has 5 columns, 4 of which overlap with `maf_df` which has 135 columns.

I'll also note that `fusion_df` is significantly smaller than `maf_df` so I am not sure what you initially meant here by _a lot_ of NAs but when running this using the subset file, the `fusion_df` data accounts for 32 of the 12147 observations in the data frame created using `bind_rows`. In other words, there are 32 rows that contain NAs.

cansavvy replied:
It might be best to be a bit more intentional about the columns being bound so that way if something fails, we get a warning kicked back. In other words, what columns do you need downstream? Can we do a `select` of those columns and then bind the `fusion_df` and `maf_df` of just those columns? This way we can also double check what information is being passed so we can construct this in a way to know if we are missing information at some point.

cansavvy commented on figures/scripts/fig3-oncoprint-landscape.R:
```diff
@@ -0,0 +1,177 @@
+coOncoplot(
+  primary_plus,
+  primary_only,
+  clinicalFeatures1 = c(
```
Looks like `clinicalFeatures1` and `clinicalFeatures2` are being fed identical vectors? If so, I'd take this vector out separately and name it something.

cansavvy commented on figures/scripts/fig3-oncoprint-landscape.R:
```diff
@@ -0,0 +1,177 @@
+root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
+
+# Path to the data obtained via `00-map-to-sample_id.R`
+scratch_oncoprint_dir <- file.path(root_dir, "scratch", "oncoprint_files")
```
Just checking, does this directory get all its necessary files from lines 53 - 76 of `generate_figures.sh`?

cbethell replied:
That is correct.

cbethell commented:
@cansavvy, I believe I implemented all of the suggestions you made above with the exception of:
>It might be best to be a bit more intentional about the columns being bound so that way if something fails, we get a warning kicked back. In other words, what columns do you need downstream? Can we do a select of those columns and then bind the fusion_df and maf_df of just those columns? This way we can also double check what information is being passed so we can construct this in a way to know if we are missing information at some point.

The reason being that when I attempted to select the variables that I believed were necessary and that were common to both data.frames, I encountered an error. So I need to do a little more digging into this to figure out what column variables are needed and which ones aren't.

In the meantime, this is ready for another look to ensure that I interpreted your other review comments correctly.

cansavvy commented on analyses/oncoprint-landscape/util/oncoplot-functions.R:
```diff
@@ -0,0 +1,77 @@
+  # Args:
+  #   maf_df: data.frame with data from a MAF file
+  #   cnv_df: data.frame with copy number variant data
+  #   metadata: data.frame with the relavant metadata
```
```suggestion
  #   metadata: data.frame with the relevant metadata
```

jaclyn-taroni commented on analyses/oncoprint-landscape/util/oncoplot-palette.R:
```diff
@@ -29,4 +29,42 @@ color_palette <- c("Missense_Mutation" = "#35978f",
-                   "Multi_Hit_Fusion" = "#CD96CD")
+                   "Multi_Hit_Fusion" = "#CD96CD",
+                   "Adenoma" = "#f23d3d",
+                   "ATRT" =	"#731d1d",
+                   "Central neurocytoma" = "#b38686",
+                   "Chondrosarcoma" =	"#cc5c33",
+                   "Chordoma" =	"#331c0d",
+                   "Choroid plexus tumor" =	"#ffb380",
+                   "CNS EFT-CIC" = "#b25f00",
+                   "CNS lymphoma"	= "#f2d6b6",
+                   "CNS neuroblastoma" = "#736556",
+                   "CNS Rhabdomyosarcoma"	= "#ffaa00",
+                   "CNS sarcoma" =	"#4c3d00",
+                   "Craniopharyngioma" =	"#e2f200",
+                   "DNET" =	"#919926",
+                   "Dysplasia" = "#d6f2b6",
+                   "Embryonal Tumor" = "#304d26",
+                   "Ependymoma" = "#00f241",
+                   "ETMR" =	"#009929",
+                   "Ganglioglioma" = "#698c7c",
+                   "Germinoma" = "#39e6c3",
+                   "Glial-neuronal tumor NOS" = "#005359",
+                   "Gliosis" =	"#263233",
+                   "Hemangioblastoma" =	"#00c2f2",
+                   "Hemangioma"	= "#40a6ff",
+                   "HGAT" =	"#406280",
+                   "Langerhans Cell histiocytosis" = "#0044ff",
+                   "LGAT" =	"#00144d",
+                   "LGMT" = "#acbbe6",
+                   "Medulloblastoma" = "#7373e6",
+                   "Meningioma" =	"#3d0099",
+                   "MPNST" = "#c200f2",
+                   "Neurofibroma" =	"#917399",
+                   "none" =	"#f1f1f1",
+                   "Oligodendroglioma" = "#f279da",
+                   "Other" = "#cc0052",
+                   "Pineoblastoma" = "#994d6b",
+                   "Schwannoma" =	"#4d2636",
+                   "Teratoma" =	"#ffbfd9"
+)
```
If the palette in `figures/palettes/histology_color_palette.tsv` were to be updated, you would have to also make changes to this code. Instead, I would find a way to use `figures/palettes/histology_color_palette.tsv` and I'd even consider moving the OncoPrint palette to a TSV files in `figures/palettes`.

jaclyn-taroni replied:
Looks like this has come up before! 😅 https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/510#issuecomment-607523137 Probably means we should do it!

jaclyn-taroni commented:
@cbethell regarding

> I was unsuccessful in my attempt to use patchwork to assemble the final figure due to what I believe is just the functionality of the oncoplot function.
It does not allow me to save the individual plots as an object in the R environment, therefore preventing me from using the patchwork::plot_layout()function.
Are there any suggestions for an alternative here?

Is there anything on #675 you can borrow?

cbethell commented:
> @cbethell regarding
> 
> > I was unsuccessful in my attempt to use patchwork to assemble the final figure due to what I believe is just the functionality of the oncoplot function.
> > It does not allow me to save the individual plots as an object in the R environment, therefore preventing me from using the patchwork::plot_layout()function.
> > Are there any suggestions for an alternative here?
> 
> Is there anything on #675 you can borrow?

I can give it a try, it looks like using the methods there to save and combine the plots into a multipanel figure should work!

cbethell commented:
@jaclyn-taroni, in this PR's latest commits, I implemented the use of the `figures/palettes/histology_color_palette.tsv` and `figures/palettes/divergent_color_palette.tsv` files to create an `figures/palettes/oncoprint_color_palette.tsv` file for use in the `oncoprint-landscape` module and in the `figures` directory. I reran the plots and added a description of this palette to `figures/README.md`. 

If the production/usage of this palette seems suffice, I can file a new PR to add the production of this palette to the `figures/scripts/color_palettes.R` script. _(Note: This removed the need for the hard coded `analyses/oncoprint-landscape/util/oncoplot-palette.R` script._

I also added the execution of the `analyses/focal-cn-file-preparation/run-prepare-cn.sh` to `figures/generate-figures.sh` upon the realization that this will be needed to produce the focal CN file that we want incorporated into the oncoprints (although this techinically will not produce the file we want until PR #686 is merged, hence the `consensus_seg_recurrent_focal_cn_units.tsv` file being added in this PR).

Additionally, as discussed, I was able to copy the final oncoprint figure from `analyses/oncoprint-landscape/plots` to `figures/pngs`, removing the need for an individual figure 3 generation script.

I also want to note that I added a rough estimate of the RAM requirements to the figure summary table of `figures/README.md`. I am not sure how far of this may be so if you have any suggestions for the best way to figure that out and make sure that estimate is accurate, please let me know!

With _ALL_ of that being said, this PR is ready for another look.

jaclyn-taroni commented on figures/README.md:
```diff
@@ -60,6 +61,7 @@ Biospecimens without a `short_histology` designation are coded as `none` and ass
 |`gradient_col_palette.tsv`| <br>gradient_0:![f7f7f7](https://placehold.it/150x40/f7f7f7/000000?text=f7f7f7) <br>gradient_1:![f7fcf5](https://placehold.it/150x40/f7fcf5/000000?text=f7fcf5) <br>gradient_2:![e5f5e0](https://placehold.it/150x40/e5f5e0/000000?text=e5f5e0) <br>gradient_3:![c7e9c0](https://placehold.it/150x40/c7e9c0/000000?text=c7e9c0) <br>gradient_4:![a1d99b](https://placehold.it/150x ...
 |`divergent_col_palette.tsv`|<br>divergent_low_5:![053061](https://placehold.it/150x40/053061/FFFFFF?text=053061) <br>divergent_low_4:![2166ac](https://placehold.it/150x40/2166ac/FFFFFF?text=2166ac) <br>divergent_low_3:![4393c3](https://placehold.it/150x40/4393c3/FFFFFF?text=4393c3) <br>divergent_low_2:![92c5de](https://placehold.it/150x40/92c5de/FFFFFF?text=92c5de) <br>divergent_low_1:![d1e5f0]( ...
 |`binary_col_palette.tsv` |<br>binary_1:![2166ac](https://placehold.it/150x40/2166ac/FFFFFF?text=2166ac) <br>binary_2:![b2182b](https://placehold.it/150x40/b2182b/FFFFFF?text=b2182b) <br>na_color:![f1f1f1](https://placehold.it/150x40/f1f1f1/000000?text=f1f1f1)|A vector of two hex codes|For binary variables e.g. presence/absence or Amp/Del as statuses|
+| `oncoprint_color_palette.tsv` | <br>Adenoma:![f23d3d](https://placehold.it/150x40/f23d3d/FFFFFF?text=f23d3d) <br>ATRT:![731d1d](https://placehold.it/150x40/731d1d/FFFFFF?text=731d1d) <br>Central neurocytoma:![b38686](https://placehold.it/150x40/b38686/FFFFFF?text=b38686) <br>Chondrosarcoma:![cc5c33](https://placehold.it/150x40/cc5c33/FFFFFF?text=cc5c33) <br>Chordoma:![331c0d](https://placehold. ...
```
This does not look like it got updated with the most recent changes to the palette.

cbethell replied:
You are correct, will do that in the upcoming commit.

jaclyn-taroni commented:
In addition to https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/686#issuecomment-632671064, I'd like to see the README for the OncoPrint module updated in a future PR. It is out of date. Since you are no longer snagging the brain cancer genes of interest in your shell script, you should document where you got them from in the README. Can you file an issue summarizing the changes that would come in future PRs @cbethell? (Alternatively, update #641.)

jaclyn-taroni commented:
I will merge once CI passes.

jaclyn-taroni commented:
This has, in fact, passed. I don't know why this still says pending.
